### PR TITLE
Evaluate custom regex in backend pwd validation.

### LINKED ISSUE #5316: Password validation inactive in BackendPassword controller
### Affected version(s)

4.13.10

### Description

As described in https://github.com/terminal42/contao-password-validation/issues/3 and addressed in https://github.com/contao/contao/pull/2190 password validation in the backend doesn't work. I offered to rebase the Pull Request in https://github.com/contao/contao/pull/2190#issuecomment-1255949902 - not sure how to proceed here.

### PR DESCRIPTION
**Pull request by AlexanderWillner**
This commit is based on the initial pull request https://github.com/contao/contao/pull/2190 that addresses https://github.com/terminal42/contao-password-validation/issues/3. In the context of https://github.com/contao/contao/issues/5316 the changes were rebased for Contao 4.13 and also refactored after initial tests.

Fixes #5316

### CODE REVIEW

File: core-bundle/src/Resources/contao/controllers/BackendPassword.php [hunk 1]
```diff
@@ -127,6 +133,42 @@ public function run()
...
+	 */
+	private function _evalRegexp($pw)
+	{
+		// no custom regex available
+		if (!isset($GLOBALS['TL_HOOKS']['addCustomRegexp']) || !\is_array($GLOBALS['TL_HOOKS']['addCustomRegexp']))
+			return false;
+
+		// load user data container
+		$this->loadDataContainer('tl_user');
+		$dc = new \DC_Table('tl_user');
```
Comment (ausi) [inline core-bundle/src/Resources/contao/controllers/BackendPassword.php]:
```suggestion
		$dc = new DC_Table('tl_user');
```

Comment (AlexanderWillner) [reply core-bundle/src/Resources/contao/controllers/BackendPassword.php]: Thanks, changed it. As https://github.com/contao/contao/actions/runs/3195462207/jobs/5216148107 states that `use Contao\DC_Table;` should not be used, I assume I should remove this line as well?

Comment (ausi) [reply core-bundle/src/Resources/contao/controllers/BackendPassword.php]: Yes.

File: core-bundle/src/Resources/contao/controllers/BackendPassword.php [hunk 2]
```diff
@@ -55,8 +55,14 @@ public function run()
...
 
+			// Make sure the password gets evaluated against custom regex if they exist
+			// First rule, as these requirements are usually stronger than the default ones
+			if ($message = self::_evalRegexp($pw))
+			{
+				Message::addError($message);
+			}
 			// Password too short
-			if (mb_strlen($pw) < Config::get('minPasswordLength'))
+			elseif (mb_strlen($pw) < Config::get('minPasswordLength'))
```
Comment (AlexanderWillner) [inline core-bundle/src/Resources/contao/controllers/BackendPassword.php:65]: to be discussed if this and other occurrences should be `else` conditions, as the user might be interested in a list of validation issues to fix them all together with a new password and not one by one...

**Comment by AlexanderWillner** [comment]
Does this change look good to you?

File: core-bundle/src/Resources/contao/controllers/BackendPassword.php [hunk 3]
```diff
@@ -127,6 +133,46 @@ public function run()
...
+	 * @param  string      $pw The password
+	 * @return bool|string The error message, false otherwise
+	 */
+	private function _evalRegexp($pw)
+	{
+		// no custom regex available
+		if (!isset($GLOBALS['TL_HOOKS']['addCustomRegexp']) || !\is_array($GLOBALS['TL_HOOKS']['addCustomRegexp']))
+		{
+			return false;
+		}
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/controllers/BackendPassword.php]: I think this needs to be removed (or changed to `empty($GLOBALS['TL_DCA']['tl_user']['fields']['password']['eval']['rgxp'])`) as `$GLOBALS['TL_DCA']['tl_user']['fields']['password']['eval']['rgxp']` can, in theory, also be one of the built-in checks.

File: core-bundle/src/Resources/contao/controllers/BackendPassword.php [hunk 4]
```diff
@@ -127,6 +133,46 @@ public function run()
...
+		{
+			$this->import($callback[0]);
+			$break = $this->{$callback[0]}->{$callback[1]}($rgxp, $pw, $widget);
+
+			// return the actual error message if it exists
+			if ($break === true && $widget->hasErrors())
+			{
+				return $widget->getErrorsAsString();
+			}
+		}
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/controllers/BackendPassword.php]: You don't need to call the hook yourself, I think you can just execute `$widget->validate()`.

**Review by fritzmg — COMMENTED** [review]
You still need to execute the CS fixer (`composer run-script cs-fixer` or at least `php tools/ecs/vendor/bin/ecs check *-bundle/src/Resources/contao --config tools/ecs/vendor/contao/easy-coding-standard/config/legacy.php --fix --ansi`).

**Comment by fritzmg** [comment]
Also, if we consider the fact that `rgxp` does not work for the back end password a bug then this should be implemented in Contao 4.9, not 4.13. If it's a feature, it needs to be based against `5.x` (for Contao 5.1).

**Comment by AlexanderWillner** [comment]
> You still need to execute the CS fixer (`composer run-script cs-fixer` or at least `php tools/ecs/vendor/bin/ecs check *-bundle/src/Resources/contao --config tools/ecs/vendor/contao/easy-coding-standard/config/legacy.php --fix --ansi`).

Thanks, however, not sure if this is the case as both, the CI (see CI results) and the actual command seem to be fine, no?

```bash
% composer run-script cs-fixer 
> @php tools/ecs/vendor/bin/ecs check *-bundle/bin *-bundle/src *-bundle/tests test-case/src tools/*/src --config tools/ecs/config/default.php --fix --ansi
 355/355 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 [OK] No errors found. Great job - your code is shiny in style!                                                         
                                                                                                                        
> @php tools/ecs/vendor/bin/ecs check *-bundle/src/Resources/contao --config tools/ecs/config/legacy.php --fix --ansi

 [OK] No errors found. Great job - your code is shiny in style!                                                         
                                                                                                                        
> @php tools/ecs/vendor/bin/ecs check *-bundle/src/Resources/contao/templates --config tools/ecs/config/template.php --fix --ansi
                                                                                                                        
 [OK] No errors found. Great job - your code is shiny in style!                                                         

% git status
On branch remotes/aw/hotfix/rgxp-backend-password
nothing to commit, working tree clean
```

File: core-bundle/src/Resources/contao/controllers/BackendPassword.php (hunk 3 above)
Comment (AlexanderWillner) [reply core-bundle/src/Resources/contao/controllers/BackendPassword.php]: lgtm, will change it in the next version (note that I'm refactoring here an existing pull request from @richardhj)

**Comment by AlexanderWillner** [comment]
> Also, if we consider the fact that `rgxp` does not work for the back end password a bug then this should be implemented in Contao 4.9, not 4.13. If it's a feature, it needs to be based against `5.x` (for Contao 5.1).

I think this is a bug as the validations works in the user management backend view, but not in this specific case. Shall I create a second PR for 4.9 after we're happy for this change in 4.13?

File: core-bundle/src/Resources/contao/controllers/BackendPassword.php (hunk 4 above)
Comment (AlexanderWillner) [reply core-bundle/src/Resources/contao/controllers/BackendPassword.php]: Indeed. The patch looks nicer now (and the tests show it still works).

**Comment by AlexanderWillner** [comment]
Does the current version look good to you?

**Comment by aschempp** [comment]
Just thinking out loud, maybe we should use the widget class to render the widget in the template in the first place?

**Comment by AlexanderWillner** [comment]
> Just thinking out loud, maybe we should use the widget class to render the widget in the template in the first place?

Well, in this case I'd need an example or guide that I could follow.

**Comment by fritzmg** [comment]
I have created an alternate version of this PR where the widget is used for validation and rendering: https://github.com/contao/contao/pull/5415

**Pull request closed without merging**